fdb-extensions/src/main/java/com/apple/foundationdb/async/hnsw/NodeKind.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * NodeKind.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2023 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.async.hnsw;
+
+import com.google.common.base.Verify;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Represents the different kinds of nodes, each associated with a unique byte value for serialization and
+ * deserialization.
+ */
+public enum NodeKind {
+    /**
+     * Compact node. Serialization and deserialization is implemented in {@link CompactNode}.
+     * <p>
+     * Compact nodes store their own vector and their neighbors-list only contain the primary key for each neighbor.
+     */
+    COMPACT((byte)0x00),
+
+    /**
+     * Inlining node. Serialization and deserialization is implemented in {@link InliningNode}.
+     * <p>
+     * Inlining nodes do not store their own vector and their neighbors-list contain the both the primary key and the
```

Comment:
Interesting that they don't store their own vector. I suppose there's some logic to it--in some way the choice is to either annotate each _vertex_ in the graph with the vector (in `COMPACT`) or to annotate each _directed edge_ in the adjacency list (with `INLINING`), and if you do the latter, there's not really a reason to do the former. The one place I think you'd need to add new information would be for the entry point into a layer of `INLINING` nodes, which otherwise wouldn't know what the vector value is.